server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -226,15 +256,7 @@ private void onFailure(int idx, String nodeId, Throwable t) {
         }
 
         private void finishHim() {
-            NodesResponse finalResponse;
-            try {
-                finalResponse = newResponse(request, responses);
-            } catch (Exception e) {
-                logger.debug("failed to combine responses from nodes", e);
-                listener.onFailure(e);
-                return;
-            }
-            listener.onResponse(finalResponse);
+            threadPool.executor(finalExecutor).execute(ActionRunnable.supply(listener, () -> newResponse(request, responses)));
```

Comment:
nit: this reintroduces double firing the listener if `listener.onResponse` fails (will also invoke `onFailure` now). I am OK with it, since it is a separate effort to come up with a good solution for the pattern you use here - unless you can see an easy way around it.

Reply:
I knew you were gonna point that out (really :)). The problem here is, that we really need this safe-guard here I think. Otherwise an exception in `listener.onResponse` (highly likely a bug in the transport code but still) would lead to the request never getting a response back leaking memory on the caller. We have the same issue when we run this on the same thread actually (the transport handler upstream brings the exception back around to the listener in that case).
-> no easy fix here for now I'm afraid. But I do think we should look into adding more assertions for this kind of thing to `ActionRunnable` like you did to `ActionListener.map` and clean these things up more for sure.